deepchem/models/jax_models/jax_model.py

```diff
@@ -0,0 +1,373 @@
+
+      if isinstance(labels, list) and len(labels) == 1:
+        labels = labels[0]
+
+      if isinstance(weights, list) and len(weights) == 1:
+        weights = weights[0]
+
+      params, opt_state, batch_loss = grad_update(params, opt_state, inputs,
+                                                  labels, weights)
+
+      avg_loss += jax.device_get(batch_loss)
+      self._global_step += 1
+      current_step = self._global_step
+      averaged_batches += 1
+      should_log = (current_step % self.log_frequency == 0)
+
+      if should_log:
+        avg_loss = float(avg_loss) / averaged_batches
+        logger.info(
+            'Ending global_step %d: Average loss %g' % (current_step, avg_loss))
+        if all_losses is not None:
+          all_losses.append(avg_loss)
+        # Capture the last avg_loss in case of return since we're resetting to 0 now
+        last_avg_loss = avg_loss
+        avg_loss = 0.0
+        averaged_batches = 0
+      for c in callbacks:
+        c(self, current_step)
+
+    # Report final results.
+    if averaged_batches > 0:
+      avg_loss = float(avg_loss) / averaged_batches
+      logger.info(
+          'Ending global_step %d: Average loss %g' % (current_step, avg_loss))
+      if all_losses is not None:
+        all_losses.append(avg_loss)
+      last_avg_loss = avg_loss
+
+    time2 = time.time()
+    logger.info("TIMING: model fitting took %0.3f s" % (time2 - time1))
+    self._set_trainable_params(params, opt_state)
+    return last_avg_loss
+
+  def _get_trainable_params(self):
+    """
+    Will be used to seperate freezing parameters while transfer learning
+    """
+    return self.params, self.opt_state
+
+  def _set_trainable_params(self, params: hk.Params, opt_state: optax.OptState):
+    """
+    A functional approach to setting the final parameters after training
+    """
+    self.params = params
+    self.opt_state = opt_state
+
+  def _create_gradient_fn(self, loss, model, optimizer, loss_outputs):
+    """
+    This function calls the update function, to implement the backpropogation
+    """
+
+    @jax.jit
+    def model_loss(params, batch, target, weights):
+      predict = model.apply(params, batch)
+      if loss_outputs is not None:
+        predict = [predict[i] for i in loss_outputs]
+      return loss(predict, target, weights)
+
+    @jax.jit
+    def update(params, opt_state, batch, target,
+               weights) -> Tuple[hk.Params, optax.OptState, jnp.ndarray]:
+      batch_loss, grads = jax.value_and_grad(model_loss)(params, batch, target,
+                                                         weights)
+      updates, opt_state = optimizer.update(grads, opt_state)
+      new_params = optax.apply_updates(params, updates)
+      return new_params, opt_state, batch_loss
+
+    return update
+
+  def _prepare_batch(self, batch):
+    inputs, labels, weights = batch
+    inputs = [
+        x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs
```

Comment:
Does mixing `jnp` and `np` matter at all?

Reply:
well `jnp` can perform all the operations that are performed in `np`, hence its not a necessary.